compiler/AST/AggregateType.cpp

```diff
@@ -2227,42 +2227,51 @@ void AggregateType::buildDefaultInitializer() {
     fn->insertFormalAtTail(_mt);
     fn->insertFormalAtTail(_this);
 
-    std::set<const char*> names;
-    SymbolMap fieldArgMap;
+    if (this->isUnion() == false) {
+      std::set<const char*> names;
+      SymbolMap fieldArgMap;
 
-    if (addSuperArgs(fn, names, fieldArgMap) == true) {
-      // Parent fields before child fields
-      fieldToArg(fn, names, fieldArgMap);
+      if (addSuperArgs(fn, names, fieldArgMap) == true) {
+        // Parent fields before child fields
+        fieldToArg(fn, names, fieldArgMap);
 
-      // Replaces field references with argument references
-      // NOTE: doesn't handle inherited fields yet!
-      update_symbols(fn, &fieldArgMap);
+        // Replaces field references with argument references
+        // NOTE: doesn't handle inherited fields yet!
+        update_symbols(fn, &fieldArgMap);
 
-      DefExpr* def = new DefExpr(fn);
+        DefExpr* def = new DefExpr(fn);
+        symbol->defPoint->insertBefore(def);
+
+        fn->setMethod(true);
+        fn->addFlag(FLAG_METHOD_PRIMARY);
+
+        preNormalizeInitMethod(fn);
+
+        normalize(fn);
+
+        // BHARSH INIT TODO: Should this be part of normalize(fn)? If we did that
+        // we would emit two use-before-def errors for classes because of the
+        // generated _new function.
+        checkUseBeforeDefs(fn);
 
```

Comment:
Looks like you removed a line here but I don't think it's worse that way (just pointing it out in case you like it better the other way)